test/E2E/Helpers.purs

```diff
@@ -0,0 +1,365 @@
+  ( hasSelector
+  , startExample
+  , exampleUrl
+  , doJQ
+  , click
+  , checkSuccess
+  , enable
+  , setAttr
+  , buttonWithText
+  , password
+  , jqStr
+  , jqSet
+  , setText
+  , setValue
+  , trigger
+  , clickButton
+  , testPassword
+  , reactSetValue
+  , retrieveJQuery
+  , showOutput
+  , inWalletPage
+  , namiConfirmAccess
+  , namiSign
+  , geroConfirmAccess
+  , geroSign
+  , geroInit
+  , Selector(..)
+  , Action(..)
+  , NoShowPage(..)
+  , RunningExample(..)
+  , E2EOutput
+  ) where
+
+import Prelude
+
+import Control.Promise (Promise, toAffE)
+import Data.Array (head, filterA)
+import Data.Foldable (intercalate)
+import Data.Maybe (Maybe(Just, Nothing), maybe)
+import Data.Newtype (class Newtype, wrap, unwrap)
+import Data.Traversable (for, fold)
+import Effect (Effect)
+import Effect.Aff (Aff, launchAff_, delay)
+import Effect.Class (liftEffect)
+import Effect.Console (log)
+import Effect.Ref (Ref)
+import Effect.Ref as Ref
+import Effect.Uncurried (mkEffectFn1, EffectFn1)
+import Foreign (Foreign, unsafeFromForeign)
+import Test.E2E.Feedback (testFeedbackIsTrue)
+import Test.Spec.Assertions (shouldSatisfy)
+import Toppokki as Toki
+
+exampleUrl :: Toki.URL
+exampleUrl = wrap "http://localhost:4008/"
+
+testPassword :: String
+testPassword = "ctlctlctl"
+
+testPasswordGero :: String
+testPasswordGero = "VZVfu5rp1r"
+
+geroSeed :: String
+geroSeed =
+  "blanket source govern pony crash genius army announce space topic fan beauty clever risk act"
+
+data OutputType = PageError | Console | RequestFailed
+
+derive instance Eq OutputType
+
+newtype E2EOutput = E2EOutput
+  { outputType :: OutputType
+  , output :: String
+  }
+
+newtype RunningExample = RunningExample
+  { browser :: Toki.Browser
+  , jQuery :: String
+  , main :: Toki.Page
+  , errors :: Ref (Array E2EOutput)
+  }
+
+derive instance Newtype RunningExample _
+
+retrieveJQuery :: Toki.Page -> Aff String
+retrieveJQuery = toAffE <<< _retrieveJQuery
+
+typeInto :: Selector -> String -> Toki.Page -> Aff Unit
+typeInto selector text page = toAffE $ _typeInto selector text page
+
+jQueryCount :: Selector -> Toki.Page -> Aff Int
+jQueryCount selector page = unsafeFromForeign <$> doJQ selector (wrap "length")
+  page
+
+hasSelector :: Selector -> Toki.Page -> Aff Boolean
+hasSelector selector page = (_ > 0) <$> jQueryCount selector page
+
+findWalletPage :: String -> Toki.Browser -> Aff (Maybe Toki.Page)
+findWalletPage jQuery browser = do
+  pages <- injectJQueryAll jQuery browser
+  pages' <- filterA (hasSelector button) pages
+  pure $ head $ pages'
+
+waitForWalletPage :: String -> Toki.Browser -> Aff Toki.Page
+waitForWalletPage jQuery browser =
+  findWalletPage jQuery browser >>= case _ of
+    Nothing -> do
+      delay $ wrap 100.0
+      waitForWalletPage jQuery browser
+    Just page -> pure page
+
+showOutput :: Ref (Array E2EOutput) -> Effect String
+showOutput ref =
+  let
+    show' (E2EOutput { outputType, output }) = showType outputType <> " " <>
```

Comment:
I think it might be easier to read if `show'` and `showType` where moved to a `where` clause with type signatures